k8s/plugin/src/resources/upgrade.rs

```diff
@@ -54,187 +62,371 @@ impl UpgradeResources {
-                        if sa.iter().count() == 0 {
-                            let ns = Some(ns.to_string());
-                            let service_account = objects::upgrade_operator_service_account(
-                                ns,
-                                uo.release_name.clone(),
-                            );
-                            let pp = PostParams::default();
-                            match uo.service_account.create(&pp, &service_account).await {
-                                Ok(sa) => {
-                                    println!(
-                                        "Service Account : {} created in namespace : {}.",
-                                        sa.metadata.name.unwrap(),
-                                        sa.metadata.namespace.unwrap(),
-                                    );
-                                }
-                                Err(e) => {
-                                    println!("Failed in creating service account  {:?}", e);
-                                    tokio::time::sleep(Duration::from_secs(1)).await;
-                                    std::process::exit(1);
-                                }
-                            }
-                        } else {
-                            println!(
-                                "Service Account : {} in namespace : {} already exist.",
-                                sa.items[0].metadata.name.as_ref().unwrap(),
-                                sa.items[0].metadata.namespace.as_ref().unwrap(),
-                            );
+    /// Create/Delete ServiceAction
+    pub async fn service_account_actions(
+        &self,
+        ns: &str,
+        action: Actions,
+    ) -> Result<(), UpgradeError> {
+        if let Some(sa) = self
+            .service_account
+            .get_opt(&upgrade_group(
+                &self.release_name,
+                UPGRADE_OPERATOR_SERVICE_ACCOUNT,
+            ))
+            .await?
+        {
+            match action {
+                Actions::Create => {
+                    println!(
+                        "ServiceAccount: {} in namespace: {} already exist.",
+                        sa.metadata.name.as_ref().unwrap(),
+                        sa.metadata.namespace.as_ref().unwrap(),
+                    );
+                }
+                Actions::Delete => {
+                    match self
+                        .service_account
+                        .delete(
+                            &upgrade_group(&self.release_name, UPGRADE_OPERATOR_SERVICE_ACCOUNT),
+                            &DeleteParams::default(),
+                        )
+                        .await
+                    {
+                        Ok(_) => {
+                            println!("ServiceAccount deleted");
+                        }
+                        Err(e) => {
+                            return Err(UpgradeError::ServiceAccountDelete { error: e });
```

Comment:
nit: if you rename e to error, the can simply have `Err(UpgradeError::ServiceAccountDelete { error })`